src/signature.rs

```diff
@@ -157,27 +210,65 @@ impl Decoded {
             // A signature is of type 0x30 (compound).
             [0x30, total_len, content @ ..] => {
                 // Maximum size constraint.
-                *total_len <= 70
-                // Make sure the length covers the entire signature.
-                    && usize::from(*total_len) == content.len()
-                    && match content {
-                        // Check whether the R element is an integer.
-                        // Extract the length of the R element.
-                        [0x02, r_len, r_s @ ..] => match r_s.split_at((*r_len).into()) {
-                            // Check whether the S element is an integer.
-                            // Extract the length of the S element.
-                            // Make sure the length of the S element is still inside the signature.
-                            (r, [0x02, s_len, s @ ..]) => {
-                                Self::is_valid_integer(r)
-                                    && usize::from(*s_len) == s.len()
-                                    && Self::is_valid_integer(s)
-                            }
-                            ([..], [..]) => false,
-                        },
-                        [..] => false,
+                if *total_len <= 70 {
+                    // Make sure the length covers the entire signature.
+                    if usize::from(*total_len) == content.len() {
+                        match content {
+                            // Check whether the R element is an integer.
+                            // Extract the length of the R element.
+                            [0x02, r_len, r_s @ ..] => match r_s.split_at((*r_len).into()) {
+                                // Check whether the S element is an integer.
+                                // Extract the length of the S element.
+                                // Make sure the length of the S element is still inside the signature.
+                                (r, [0x02, s_len, s @ ..]) => Self::is_valid_integer(r)
```

Comment:
If you pass the name and length into `Self.is_valid_integer` (maybe call it `parse_integer`), then it can return a `Result` itself, and I think in that case you can deduplicate the error handling between `r` and `s`. (The length of `r_s` will be checked redundantly but it doesn't matter.)